Test get subnet

majorchork opened the pull request:
<!--
Please add the affected binary name in the title unless multiple binaries are affected, e.g.
[cinder-csi-plugin] Fix volume deletion
For openstack-cloud-controller-manager, you can use [occm] for short.

All the currently maintained binaries are:
* openstack-cloud-controller-manager (occm)
* cinder-csi-plugin
* manila-csi-plugin
* k8s-keystone-auth
* client-keystone-auth
* octavia-ingress-controller
* magnum-auto-healer
* barbican-kms-plugin
-->

**What this PR does / why we need it**:
this pr adds unit test for getSubnetID LbaasV2 method in loadbalancer.go
**Which issue this PR fixes(if applicable)**:
refers #2400 

**Special notes for reviewers**:
<!-- e.g. How to test this PR -->

**Release note**:
<!--
1. Release note is required if a significant change is introduced, otherwise please keep this section as is.
2. Release note is in Markdown format and should begin with the binary name unless multiple binaries are affected, e.g. [openstack-cloud-controller-manager] Deprecate Neutron-LBaaS support.
3. Instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

dulek commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -787,3 +788,149 @@ func Test_buildPoolCreateOpt(t *testing.T) {
+			expectedErr: nil,
+		},
+		{
+			name: "test get subnet from config class",
```
This name is the same as in line 841, we should have them unique.

dulek commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -787,3 +788,149 @@ func Test_buildPoolCreateOpt(t *testing.T) {
+				assert.EqualError(t, err, tt.expectedErr.Error())
+			}
+
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getSubnetID() = %v, expected %v", got, tt.want)
+			}
```
Why not `assert.Equal()`?

dulek replied:
I think this one still stands.

majorchork replied:
@dulek updated

dulek reviewed (commented):
Seems like a case testing [this codepath](https://github.com/kubernetes/cloud-provider-openstack/blob/8af230a0d5b5443ef405017a3c541a9556ee0158/pkg/openstack/loadbalancer.go#L1467-L1470) is missing?

dulek commented:
/ok-to-test

majorchork commented:
> Seems like a case testing [this codepath](https://github.com/kubernetes/cloud-provider-openstack/blob/8af230a0d5b5443ef405017a3c541a9556ee0158/pkg/openstack/loadbalancer.go#L1467-L1470) is missing?

yea! that was the one with the wrong name, I've updated it now

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -593,3 +595,296 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"test-class": {
+									SubnetID:           "test-id",
```
What do you think about putting something like `test-class-subnet-id` here, to make sure that the subnet ID is actually taken from the annotation?

majorchork replied:
sound great! i've added it in

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -593,3 +595,296 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
```
What do you think about adding one LBClass here that is NOT `test-class`?

majorchork replied:
That sounds great, should I add an extra LBClass or completely replace the other LBClass?

majorchork replied:
i've added an extra LBClass, if i completely remove it the function will set LBClass to "" and pick the default class instead

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -593,3 +595,296 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							SubnetID: "test-default-subnet-id",
```
I'd add a LBClass here as well, just to make sure it's ignored in case the class is not set below. What do you think?

majorchork replied:
sounds nice but looking at the function the default config is only selected when the config class is empty, so if I add the extra class, it's supposed to pick it up and not ignore it

majorchork replied:
if you look at the screenshot, if the configClassName is not set to ""(empty string) the default config class is not picked, if I add a LbClass here the default config class won't be ignored, and we're testing for the default config class, if you'd like I can add an extra test case with LbClass set and the default  config class set, it should ignore the default config class and pick the LBClass
<img width="725" alt="Screenshot 2023-10-10 at 13 27 20" src="https://github.com/kubernetes/cloud-provider-openstack/assets/69452689/eee838d6-0a20-434e-88b2-91413ccc1285">

majorchork replied:
i misunderstood you before, I've added it in now

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -593,3 +595,296 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.args.lbaasV2.getSubnetID(tt.args.service, tt.args.svcConf)
+			if tt.expectedErr != nil {
```
Watch out! You are not checking the case where `tt.expectedErr` is nil and `err` is not nil

majorchork replied:
yea!! that's what I'm referring to as well, checking it is repetitive since the assert checks it, as shown in this screenshot
<img width="750" alt="Screenshot 2023-10-10 at 13 17 47" src="https://github.com/kubernetes/cloud-provider-openstack/assets/69452689/aae57cff-fef8-47f7-af84-f6b0b668b28a">

pierreprinetti replied:
Am I blind? I think your code only checks err if tt.expectedErr is not nil.

pierreprinetti reviewed (commented):
Reviewing `Test_getSubnetID`

pierreprinetti commented:
/assign

majorchork commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -824,3 +825,151 @@ func Test_buildPoolCreateOpt(t *testing.T) {
+				assert.EqualError(t, err, tt.expectedErr)
+			}
+
+			assert.Equal(t, got, tt.want)
```
Sure! I've added in the check. Please review again

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -870,3 +871,151 @@ func Test_getSecurityGroupName(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.args.lbaasV2.getSubnetID(tt.args.service, tt.args.svcConf)
+			if tt.expectedErr != "" || err != nil {
```
This would indeed work in all likely cases. However, it is not quite correct.

`assert.EqualError` is designed for cases where an error is expected, and only match its message against a string.

Now consider this edge case:
```Go
func TestAssertEqualError(t *testing.T) {
	t.Run("edge case: non-nil error with empty message", func(t *testing.T) {
		var (
			err         error  = fmt.Errorf("")
			expectedErr string = ""
		)
		assert.EqualError(t, err, expectedErr) // PASS: the error message matches the expected string
	})
}
```
In this example the error is not nil, however the assertion would not trigger any test failure.

Were the function under test to return a non-nil error with an empty string as its message, this test would not catch the issue.

Clearly, we are unlikely to hit this particular case. However, by the law of large numbers, an issue like this WILL happen, and those are precisely the hardest to debug. Let's be a little defensive in our tests, at the price of a couple more lines of code.

Here you probably should write a separate block for when `tt.expectedErr` is the empty string, because in that case we really want to `assert.NoError(t, err)`.

majorchork replied:
wow!! nice catch i really didn't consider that edge case! thanks a lot for the insight, I'll fix it

dulek commented:
/approve

I'll leave `/lgtm` for @pierreprinetti, I'm not 100% sure this is what he meant on the error checking front.

majorchork commented:
> /approve
> 
> I'll leave `/lgtm` for @pierreprinetti, I'm not 100% sure this is what he meant on the error checking front.

Thanks a lot! Noted

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -973,3 +974,129 @@ func TestLbaasV2_updateServiceAnnotations(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.args.lbaasV2.getSubnetID(tt.args.service, tt.args.svcConf)
+			if tt.expectedErr != "" || err != nil {
```
```suggestion
			if tt.expectedErr != "" {
```

pierreprinetti commented on pkg/openstack/loadbalancer_test.go:
```diff
@@ -1629,3 +1630,129 @@ func TestLbaasV2_getMemberSubnetID(t *testing.T) {
+			if tt.expectedErr == "" {
+				assert.NoError(t, err)
+			}
+			assert.Equal(t, got, tt.want)
```
I know it's very confusing, but the argument order in `assert.Equal`'s [is the other way around](https://github.com/stretchr/testify/blob/4ae48e988c8c84c138d688eba40bfc869f4be907/assert/assertions.go#L415C18-L415C18). This is important to get a clear error message in case of mismatch.

```suggestion
			assert.Equal(t, tt.want, got)
```

majorchork replied:
I agree, nice catch! very true! thanks for always being thorough! 🙏🏾

pierreprinetti replied:
thorough is my middle name

pierreprinetti replied:
![shameless](https://i.imgflip.com/13se6u.jpg)

majorchork replied:
😂😂😂 perks of the job!

pierreprinetti reviewed (approved):
/lgtm

pierreprinetti commented:
Please rebase to enable the merge :pray:

majorchork commented:
> Please rebase to enable the merge 🙏

I already have seems like the bot forgot to remove the label, I'll just force push again and see if the bot picks it this time
<img width="794" alt="Screenshot 2023-10-19 at 06 43 21" src="https://github.com/kubernetes/cloud-provider-openstack/assets/69452689/301399dd-5f49-491a-84a1-85e27be1643a">

majorchork commented:
@k8s-ci-robot  verify that this pr need rebase

pierreprinetti commented:
> > Please rebase to enable the merge 🙏
> 
> I already have seems like the bot forgot to remove the label, I'll just force push again and see if the bot picks it this time <img alt="Screenshot 2023-10-19 at 06 43 21" width="794" src="https://user-images.githubusercontent.com/69452689/276583363-301399dd-5f49-491a-84a1-85e27be1643a.png">

What do you mean, "Your branch is ahead of 'origin/master' by 17 commits"? This is not expected. Your local `master` branch is not aligned with origin.

To rebase on origin's master, try:

```sh
git checkout test-getSubnetID
git fetch origin
git rebase origin/master
```

---

Additionally, if you want to get rid of those 17 commits on your local `master` branch, here's how to do it:
```sh
git checkout master
git fetch origin
git reset --hard origin/master
git clean -fdx
```

majorchork commented:
> ```shell
> git reset --hard origin/master
> ```
<img width="932" alt="Screenshot 2023-10-19 at 07 13 47" src="https://github.com/kubernetes/cloud-provider-openstack/assets/69452689/b6061374-a555-4148-b385-f1dc5fbd47a9">


😅 my bad, it's ahead cause I've not published my local to my origin, I always just fetch upstream and merge then checkout to the branch I want to rebase before pushing, I never actually push the changes on my local origin
`git fetch upstream (upstream is  https://github.com/kubernetes/cloud-provider-openstack/master)
 git merge upstream/master
 git checkout test-getSubnetId
 git rebase master`
so my local master is ahead by all the changes I've fetched from the upstream, 
I've updated, gotten rid of the commits, and rebased now

pierreprinetti commented:
/lgtm
